File: js/src/core/event-log.ts
import { Buffer } from 'buffer';
import * as v from 'valibot';
import {
  type DstackEventLogEntrySchema,
  DstackEventLogSchema,
  SerializedDstackEventLogSchema,
} from '../schemas';
import type { AttestationEventLog } from '../types/attestation-common';
import type { RuntimeMeasurements } from '../types/verification';
import { VerificationError } from '../utils/errors';
import { sha384, trimHexPrefix, utf8 } from '../utils/common';

// dstack reserves this event type for runtime payloads, whose digest must be
// recomputed from the payload rather than trusted from the event log.
const DSTACK_RUNTIME_EVENT_TYPE = 0x08000001;

type EventLogEntry = v.InferOutput<typeof DstackEventLogEntrySchema> & {
  path: string;
};
type DecodeEventHexParams = {
  value: string;
  path: string;
  allowEmpty: boolean;
};

/**
 * Replay RTMR3 from the dstack event log. Runtime event payloads are hashed
 * again before extending the register, preventing payload substitution with a
 * copied digest. Other event types remain replayable for compatibility, but
 * cannot supply runtime metadata because their name and payload are not bound
 * by this verifier.
 */
export async function verifyAndReplayRtmr3(
  eventLog: AttestationEventLog,
  quotedRtmr3: Uint8Array,
): Promise<RuntimeMeasurements> {
  const events = parseEventLog(eventLog);
  const expected = Buffer.from(quotedRtmr3);
  if (expected.length !== 48) {
    throw new VerificationError({
      code: 'measurement.rtmr3_mismatch',
      details: {
        reason: 'wrong_length',
        expectedBytes: 48,
        actualBytes: expected.length,
      },
    });
  }

  let replayed: Uint8Array = Buffer.alloc(48);
  let count = 0;
  let osImageHash: string | undefined;
  let composeHash: string | undefined;

  for (const entry of events) {
    if (entry.imr !== 3) {
      continue;
    }
    count += 1;
    const digest = await eventDigest(entry);
    replayed = await sha384(Buffer.concat([replayed, digest]));

    // Only dstack runtime events authenticate the event name and payload: the
    // digest is recomputed from both values in eventDigest(). A legacy or
    // other event type contributes to RTMR3 replay, but its JSON metadata is
    // not authenticated and must not be exposed as a measured value.
    if (entry.event_type === DSTACK_RUNTIME_EVENT_TYPE) {
      if (entry.event === 'os-image-hash') {
        osImageHash = entry.event_payload;
      } else if (entry.event === 'compose-hash') {
        composeHash = entry.event_payload;
      }
    }
  }

  if (count === 0) {
    throw new VerificationError({
      code: 'measurement.rtmr3_mismatch',
      details: { reason: 'no_events' },
    });
  }
  if (!Buffer.from(replayed).equals(expected)) {
    throw new VerificationError({
      code: 'measurement.rtmr3_mismatch',
      details: { reason: 'replay_mismatch' },
    });
  }

  return { osImageHash, composeHash };
}

function parseEventLog(eventLog: AttestationEventLog): EventLogEntry[] {
  const schema =
    typeof eventLog === 'string'
      ? SerializedDstackEventLogSchema
      : DstackEventLogSchema;
  const parsed = v.safeParse(schema, eventLog);
  if (!parsed.success) {
    throw invalidEventLogSchema(parsed.issues[0]);
  }

  return parsed.output.map((entry, index) => ({
    ...entry,
    path: `eventLog[${index}]`,
  }));
}

function invalidEventLogSchema(issue: v.BaseIssue<unknown>): VerificationError {
  if (issue.type === 'parse_json') {
    return new VerificationError({
      code: 'measurement.event_log_invalid',
      details: { path: 'eventLog', reason: 'invalid_json' },
    });
  }

  const dotPath = v.getDotPath(issue);
  const path = !dotPath
    ? 'eventLog'
    : dotPath.startsWith('[')
      ? `eventLog${dotPath}`
      : `eventLog.${dotPath}`.replace(/\.(\d+)(?=\.|$)/g, '[$1]');
  return new VerificationError({
    code: 'measurement.event_log_invalid',
    details: {
      path,
      reason: 'invalid_type',
      expected: issue.expected ?? 'valid event log',
    },
  });
}

async function eventDigest(entry: EventLogEntry): Promise<Buffer> {
  if (entry.event_type === DSTACK_RUNTIME_EVENT_TYPE) {
    const payload = decodeEventHex({
      value: entry.event_payload,
      path: `${entry.path}.event_payload`,
      allowEmpty: true,
    });
    const eventType = Buffer.alloc(4);
    eventType.writeUInt32LE(DSTACK_RUNTIME_EVENT_TYPE);
    const computed = await sha384(
      Buffer.concat([
        eventType,
        utf8(':'),
        utf8(entry.event),
        utf8(':'),
        payload,
      ]),
    );

    if (entry.digest.length > 0) {
      const stored = decodeEventHex({
        value: entry.digest,
        path: `${entry.path}.digest`,
        allowEmpty: false,
      });
      if (stored.length !== 48 || !stored.equals(computed)) {
        throw new VerificationError({
          code: 'measurement.event_log_invalid',
          details: {
            path: `${entry.path}.digest`,
            reason: stored.length === 48 ? 'digest_mismatch' : 'wrong_length',
            ...(stored.length === 48
              ? {}
              : { expectedBytes: 48, actualBytes: stored.length }),
          },
        });
      }
    }
    return computed;
  }

  // dstack event-log digests are SHA-384 values, so a malformed length is
  // rejected even though replaying a shorter byte string would be possible.
  const digest = decodeEventHex({
    value: entry.digest,
    path: `${entry.path}.digest`,
    allowEmpty: false,
  });
  if (digest.length !== 48) {
    throw new VerificationError({
      code: 'measurement.event_log_invalid',
      details: {
        path: `${entry.path}.digest`,
        reason: 'wrong_length',
        expectedBytes: 48,
        actualBytes: digest.length,
      },
    });
  }
  return digest;
}

/** dstack runtime events may legitimately carry an empty payload. */
function decodeEventHex({
  value,
  path,
  allowEmpty,
}: DecodeEventHexParams): Buffer {
  const normalized = trimHexPrefix(value);
  if (
    (!allowEmpty && normalized.length === 0) ||
    normalized.length % 2 !== 0 ||
    !/^[0-9a-fA-F]*$/.test(normalized)
  ) {
    throw new VerificationError({
      code: 'measurement.event_log_invalid',
      details: { path, reason: 'invalid_hex' },
    });
  }
  return Buffer.from(normalized, 'hex');
}

File: py/src/verifiable_ai_sdk/core/event_log.py
"""dstack RTMR3 event-log replay."""

from __future__ import annotations

from pydantic import ValidationError

from ..schemas import DstackEventLogEntrySchema, DstackEventLogSchema
from ..types.attestation_common import AttestationEventLog
from ..types.verification import RuntimeMeasurements
from ..utils.common import sha384, trim_hex_prefix
from ..utils.errors import verification_failure


DSTACK_RUNTIME_EVENT_TYPE = 0x08000001

EVENT_LOG_FIELD_EXPECTATIONS = {
    'digest': 'string',
    'event': 'string',
    'event_payload': 'string',
    'imr': 'unsigned 32-bit integer',
    'event_type': 'unsigned 32-bit integer',
}


def verify_and_replay_rtmr3(
    event_log: AttestationEventLog, quoted_rtmr3: bytes
) -> RuntimeMeasurements:
    events = _parse_event_log(event_log)
    if len(quoted_rtmr3) != 48:
        raise verification_failure(
            'measurement.rtmr3_mismatch',
            {
                'reason': 'wrong_length',
                'expectedBytes': 48,
                'actualBytes': len(quoted_rtmr3),
            },
        )

    replayed = bytes(48)
    count = 0
    os_image_hash: str | None = None
    compose_hash: str | None = None
    for index, entry in enumerate(events):
        if entry.imr != 3:
            continue
        count += 1
        path = f'eventLog[{index}]'
        replayed = sha384(replayed + _event_digest(entry, path))
        if entry.event_type == DSTACK_RUNTIME_EVENT_TYPE:
            if entry.event == 'os-image-hash':
                os_image_hash = entry.event_payload
            elif entry.event == 'compose-hash':
                compose_hash = entry.event_payload

    if count == 0:
        raise verification_failure(
            'measurement.rtmr3_mismatch', {'reason': 'no_events'}
        )
    if replayed != quoted_rtmr3:
        raise verification_failure(
            'measurement.rtmr3_mismatch',
            {'reason': 'replay_mismatch'},
        )
    return RuntimeMeasurements(
        os_image_hash=os_image_hash,
        compose_hash=compose_hash,
    )


def _parse_event_log(event_log: AttestationEventLog) -> list[DstackEventLogEntrySchema]:
    try:
        if isinstance(event_log, str):
            return DstackEventLogSchema.model_validate_json(event_log).root
        return DstackEventLogSchema.model_validate(event_log).root
    except ValidationError as error:
        raise _invalid_event_log_schema(error) from error


def _invalid_event_log_schema(error: ValidationError):
    issue = error.errors(include_url=False)[0]
    if issue['type'] == 'json_invalid':
        return _invalid_event('eventLog', 'invalid_json', cause=error)

    location = issue['loc']
    path = _event_log_error_path(location)
    expected = _event_log_expected(location)
    return _invalid_event(path, 'invalid_type', expected=expected, cause=error)


def _event_log_error_path(location: tuple[int | str, ...]) -> str:
    return 'eventLog' + ''.join(
        f'[{part}]' if isinstance(part, int) else f'.{part}' for part in location
    )


def _event_log_expected(location: tuple[int | str, ...]) -> str:
    if not location:
        return 'array'
    field = location[-1]
    if isinstance(field, str):
        return EVENT_LOG_FIELD_EXPECTATIONS.get(field, 'valid event log')
    return 'object'


def _event_digest(entry: DstackEventLogEntrySchema, path: str) -> bytes:
    if entry.event_type == DSTACK_RUNTIME_EVENT_TYPE:
        payload = _decode_event_hex(
            entry.event_payload, f'{path}.event_payload', allow_empty=True
        )
        computed = sha384(
            DSTACK_RUNTIME_EVENT_TYPE.to_bytes(4, 'little')
            + b':'
            + entry.event.encode('utf-8')
            + b':'
            + payload
        )
        if entry.digest:
            stored = _decode_event_hex(entry.digest, f'{path}.digest')
            if len(stored) != 48 or stored != computed:
                details: dict[str, object] = {
                    'path': f'{path}.digest',
                    'reason': 'digest_mismatch'
                    if len(stored) == 48
                    else 'wrong_length',
                }
                if len(stored) != 48:
                    details.update({'expectedBytes': 48, 'actualBytes': len(stored)})
                raise verification_failure('measurement.event_log_invalid', details)
        return computed

    digest = _decode_event_hex(entry.digest, f'{path}.digest')
    if len(digest) != 48:
        raise verification_failure(
            'measurement.event_log_invalid',
            {
                'path': f'{path}.digest',
                'reason': 'wrong_length',
                'expectedBytes': 48,
                'actualBytes': len(digest),
            },
        )
    return digest


def _decode_event_hex(value: str, path: str, *, allow_empty: bool = False) -> bytes:
    normalized = trim_hex_prefix(value)
    if (
        (not allow_empty and not normalized)
        or len(normalized) % 2 != 0
        or any(character not in '0123456789abcdefABCDEF' for character in normalized)
    ):
        raise _invalid_event(path, 'invalid_hex')
    return bytes.fromhex(normalized)


def _invalid_event(
    path: str,
    reason: str,
    *,
    expected: str | None = None,
    cause: BaseException | None = None,
):
    details: dict[str, object] = {'path': path, 'reason': reason}
    if expected is not None:
        details['expected'] = expected
    return verification_failure('measurement.event_log_invalid', details, cause=cause)
